hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/FlinkInternalSchemaManager.java

```diff
@@ -0,0 +1,145 @@
+ * If type of field is changed, its old type will be returned, and projection will be created that will convert the old type to the queried one.
+ */
+public final class FlinkInternalSchemaManager implements Serializable {
+  private static final long serialVersionUID = 1L;
+
+  private final HoodieTableMetaClient metaClient;
+  private final InternalSchema querySchema;
+
+  /**
+   * Creates the manager if schema evolution enabled.
+   */
+  public static Option<FlinkInternalSchemaManager> of(Configuration conf) {
+    if (conf.getBoolean(FlinkOptions.SCHEMA_EVOLUTION_ENABLED)) {
+      HoodieTableMetaClient metaClient = StreamerUtil.createMetaClient(conf);
+      return new TableSchemaResolver(metaClient)
+          .getTableInternalSchemaFromCommitMetadata()
+          .map(schema -> new FlinkInternalSchemaManager(metaClient, schema));
+    } else {
+      return Option.empty();
+    }
+  }
+
+  FlinkInternalSchemaManager(HoodieTableMetaClient metaClient, InternalSchema querySchema) {
+    this.metaClient = metaClient;
+    this.querySchema = querySchema;
+  }
+
+  /**
+   * Returns query schema as InternalSchema.
+   */
+  public InternalSchema getQuerySchema() {
+    return querySchema;
+  }
+
+  /**
+   * Returns schema of fileSplit.
+   */
+  public InternalSchema getActualSchema(FileInputSplit fileSplit) {
+    return getActualSchema(FSUtils.getCommitTime(fileSplit.getPath().getName()));
+  }
+
+  /**
+   * Returns schema of mor fileSplit.
+   */
+  public InternalSchema getActualSchema(MergeOnReadInputSplit split) {
+    Option<String> basePath = split.getBasePath();
+    String commitTime;
+    if (basePath.isPresent()) {
+      String name = new Path(basePath.get()).getName();
+      commitTime = FSUtils.getCommitTime(name);
+    } else {
+      commitTime = split.getLatestCommit();
+    }
+    return getActualSchema(commitTime);
+  }
+
+  /**
+   * Returns list of field names in internalSchema.
+   */
+  public List<String> getFieldNames(InternalSchema internalSchema) {
+    return internalSchema.columns().stream().map(Types.Field::name).collect(Collectors.toList());
+  }
+
+  /**
+   * Returns list of field types in internalSchema.
+   */
+  public List<DataType> getFieldTypes(InternalSchema internalSchema) {
+    return AvroSchemaConverter.convertToDataType(
+        AvroInternalSchemaConverter.convert(internalSchema, getTableName())).getChildren();
+  }
+
+  /**
+   * Returns castMap based on conversions from actualSchema to querySchema.
+   */
+  public CastMap getCastMap(InternalSchema querySchema, InternalSchema actualSchema) {
+    return CastMap.of(getTableName(), querySchema, actualSchema);
+  }
+
+  /**
+   * Returns array of types positioned in fieldTypes according to selectedFields.
+   */
+  public LogicalType[] project(List<DataType> fieldTypes, int[] selectedFields) {
+    return Arrays.stream(selectedFields)
```

Comment:
The method does not belong here, may put in `DataTypeUtils`.

Reply:
Method removed